### PR TITLE
executor: fix IsPointGet judgment condition

### PR DESCRIPTION
**Pull request by tiancaiamao**
### What problem does this PR solve? 

index lookup should not use the max ts optimization

### What is changed and how it works?

fix IsPointGet judgment condition

### Check List 

Tests 

 - Unit test

Related changes

 - Need to cherry-pick to the release branch

### CODE REVIEW

**Comment by shenli** [comment]
@tiancaiamao Please fix CI.

File: executor/adapter.go [hunk 1]
```diff
@@ -500,14 +500,15 @@ func IsPointGetWithPKOrUniqueKeyByAutoCommit(ctx sessionctx.Context, p plannerco
...
 	case *plannercore.PhysicalTableReader:
 		tableScan := v.TablePlans[0].(*plannercore.PhysicalTableScan)
 		return len(tableScan.Ranges) == 1 && tableScan.Ranges[0].IsPoint(ctx.GetSessionVars().StmtCtx), nil
 	case *plannercore.PointGetPlan:
-		return true, nil
+		// If the PointGetPlan need to read data using unique index (double read), we
+		// can't use max uint64, because using math.MaxUint64 can't guarantee repeatable-read
+		// and the data and index would be inconsistent!
+		return v.IndexInfo == nil, nil
+		// return true, nil
```
Comment (jackysp) [inline executor/adapter.go]: remove this line?

**Comment by tiancaiamao** [comment]
PTAL @jackysp @winkyao @shenli

File: go.mod [hunk 2]
```diff
@@ -55,6 +55,7 @@ require (
 	github.com/spaolacci/murmur3 v0.0.0-20180118202830-f09979ecbc72
 	github.com/struCoder/pidusage v0.1.2
 	github.com/tiancaiamao/appdash v0.0.0-20181126055449-889f96f722a2
+	github.com/tiancaiamao/debugger v0.0.0-20190426085344-91156e93f38e
```
Comment (winoros) [inline go.mod]: Can we put this package into `failpoint`?

File: executor/adapter.go [hunk 3]
```diff
@@ -500,14 +500,14 @@ func IsPointGetWithPKOrUniqueKeyByAutoCommit(ctx sessionctx.Context, p plannerco
...
 		return indexScan.IsPointGetByUniqueKey(ctx.GetSessionVars().StmtCtx), nil
-	case *plannercore.PhysicalIndexLookUpReader:
-		indexScan := v.IndexPlans[0].(*plannercore.PhysicalIndexScan)
-		return indexScan.IsPointGetByUniqueKey(ctx.GetSessionVars().StmtCtx), nil
 	case *plannercore.PhysicalTableReader:
 		tableScan := v.TablePlans[0].(*plannercore.PhysicalTableScan)
 		return len(tableScan.Ranges) == 1 && tableScan.Ranges[0].IsPoint(ctx.GetSessionVars().StmtCtx), nil
 	case *plannercore.PointGetPlan:
-		return true, nil
+		// If the PointGetPlan need to read data using unique index (double read), we
```
Comment (winoros) [inline executor/adapter.go]:
```suggestion
		// If the PointGetPlan needs to read data using unique index (double read), we
```

File: executor/executor_test.go [hunk 4]
```diff
@@ -66,6 +67,7 @@ import (
 	"github.com/pingcap/tidb/util/testutil"
 	"github.com/pingcap/tidb/util/timeutil"
 	"github.com/pingcap/tipb/go-tipb"
+	"github.com/tiancaiamao/debugger"
```
Comment (winkyao) [inline executor/executor_test.go]: Could we remove this package to tidb？

File: executor/executor_test.go [hunk 5]
```diff
@@ -1872,6 +1875,34 @@ func (s *testSuite) TestIsPointGet(c *C) {
...
+	tk1 := testkit.NewTestKit(c, s.store)
+	tk1.MustExec("use test")
+	tk1.MustExec(`create table point_get (a int, b int, c int,
+			primary key k_a(a),
+			unique key k_b(b))`)
+	tk1.MustExec("insert into point_get values (1, 1, 1)")
+	tk2 := testkit.NewTestKit(c, s.store)
+	tk2.MustExec("use test")
+
+	ctx := context.WithValue(context.Background(), "pointGetRepeatableReadTest", true)
```
Comment (winkyao) [inline executor/executor_test.go]: you can use failpoint.WithHook, then you can remove `ctx.Value("pointGetRepeatableReadTest") != nil` check:

```
	failpoint.Inject("pointGetRepeatableReadTest", func(val failpoint.Value) {
			if val.(bool) && ctx.Value("pointGetRepeatableReadTest") != nil {
				label := debugger.Bind("point-get-g1")
				debugger.Break(label)
			}
		})
```

File: executor/point_get.go [hunk 6]
```diff
@@ -88,6 +90,14 @@ func (e *PointGetExecutor) Next(ctx context.Context, req *chunk.RecordBatch) err
 		if err1 != nil {
 			return err1
 		}
+
+		failpoint.Inject("pointGetRepeatableReadTest", func(val failpoint.Value) {
+			if val.(bool) && ctx.Value("pointGetRepeatableReadTest") != nil {
+				label := debugger.Bind("point-get-g1")
+				debugger.Break(label)
```
Comment (winkyao) [inline executor/point_get.go]: Maybe `Breakpoint` is better?

File: executor/executor_test.go [hunk 7]
```diff
@@ -1872,6 +1875,34 @@ func (s *testSuite) TestIsPointGet(c *C) {
...
+		c.Assert(err, IsNil)
+		result := tk1.ResultSetToResultWithCtx(ctx, rs[0], Commentf("execute sql fail"))
+		result.Check(testkit.Rows("1"))
+	}()
+
+	label := debugger.Bind("point-get-g2")
+	debugger.Continue("point-get-g1")
+	debugger.Break(label)
+	tk2.MustExec("update point_get set b = 2, c = 2 where a = 1")
+	debugger.Continue("point-get-g1")
```
Comment (winkyao) [inline executor/executor_test.go]: continue "point-get-g2"?

File: executor/adapter.go [hunk 8]
```diff
@@ -500,14 +500,14 @@ func IsPointGetWithPKOrUniqueKeyByAutoCommit(ctx sessionctx.Context, p plannerco
...
+		// can't use max uint64, because using math.MaxUint64 can't guarantee repeatable-read
+		// and the data and index would be inconsistent!
+		return v.IndexInfo == nil, nil
```
Comment (winkyao) [inline executor/adapter.go]: https://github.com/pingcap/tidb/pull/10278/files#diff-350127760839dbfd52d23927f7ff2d95R319 can we rename the local variable from `isPointGet` to `maxTsOptimizedOpr`, or something else? isPointGet is very confused.

File: executor/executor_test.go (hunk 5 above)
Comment (tiancaiamao) [reply executor/executor_test.go]: Done

File: executor/executor_test.go (hunk 7 above)
Comment (tiancaiamao) [reply executor/executor_test.go]: continue "point-get-g1" is right here...  it means notify the other goroutine to continiue, not this one...
Notify the `"select c from point_get where b = 1"` transaction, to continue the second get(key)

File: executor/adapter.go (hunk 8 above)
Comment (tiancaiamao) [reply executor/adapter.go]: The comment is informative enough, and I also rename here
https://github.com/pingcap/tidb/pull/10278/files#diff-350127760839dbfd52d23927f7ff2d95R319

File: executor/executor_test.go (hunk 4 above)
Comment (tiancaiamao) [reply executor/executor_test.go]: IMHO, it's better for this PR to focus on bug fix only.
Maybe I could clean up here and move the package to TiDB in a seperate PR.

File: executor/point_get.go (hunk 6 above)
Comment (tiancaiamao) [reply executor/point_get.go]: Done

File: go.mod (hunk 2 above)
Comment (tiancaiamao) [reply go.mod]: https://github.com/pingcap/tidb/pull/10278#discussion_r279179951

**Comment by winkyao** [comment]
Please fix ci

**Comment by tiancaiamao** [comment]
I tried `failpoint.InjectWithContext` but with no luck, so I reset the last commit. @winkyao

**Comment by winkyao** [comment]
@tiancaiamao well, let's focus on the bug fixing.

**Pull request merged**